vm/interpreter/src/vm.rs

```diff
@@ -172,16 +174,23 @@ where
     /// Flushes the StateTree and perform a state migration if there is a migration at this epoch.
     /// If there is no migration this function will return Ok(None).
     #[allow(unreachable_code, unused_variables)]
-    pub fn migrate_state(&mut self, epoch: ChainEpoch) -> Result<Option<Cid>, Box<dyn StdError>> {
+    pub fn migrate_state(
+        &mut self,
+        epoch: ChainEpoch,
+        arc_store: Arc<impl BlockStore + Send + Sync>,
+    ) -> Result<Option<Cid>, Box<dyn StdError>> {
         match epoch {
-            x if x == UPGRADE_PLACEHOLDER_HEIGHT => {
+            x if x == UPGRADE_ACTORS_V4_HEIGHT => {
+                let start = std::time::Instant::now();
+                log::info!("Running actors_v4 state migration");
                 // need to flush since we run_cron before the migration
                 let prev_state = self.flush()?;
-                // new_state is new state root we can from calling the migration function
-                // that will be provided from the actors create (probably).
-                // TODO perform migration here, currently skipping
-                let new_state: Cid = prev_state;
+                let new_state = run_nv12_migration(arc_store, prev_state, epoch)?;
                 if new_state != prev_state {
+                    log::info!(
+                        "actors_v4 state migration successful, took: {}ms",
+                        start.elapsed().as_millis()
+                    );
                     Ok(Some(new_state))
                 } else {
                     Ok(None)
```

Comment:
Curious why this doesn't result in an error

Reply:
Seems like the state roots _must not_ match after the migration

Reply:
Right, refactored to return an error now